[MacOS] Implement scrollbar-color

nmoucht opened the pull request:
#### 9efb2f30da2dce7d150cbd05eb67e94330d65388
<pre>
[MacOS] Implement scrollbar-color
<a href="https://bugs.webkit.org/show_bug.cgi?id=297166">https://bugs.webkit.org/show_bug.cgi?id=297166</a>
<a href="https://rdar.apple.com/157906557">rdar://157906557</a>

Reviewed by Tim Nguyen and Simon Fraser.

Adopt new AppKit SPI for setting the color of the scrollbar thumb and track.

* LayoutTests/fast/scrolling/mac/scrollbars/scrollbar-color-body-expected.txt: Added.
* LayoutTests/fast/scrolling/mac/scrollbars/scrollbar-color-body.html: Added.
* LayoutTests/fast/scrolling/mac/scrollbars/scrollbar-color-expected.txt: Added.
* LayoutTests/fast/scrolling/mac/scrollbars/scrollbar-color.html: Added.
* LayoutTests/platform/mac-sequoia/TestExpectations:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/PAL/pal/spi/mac/NSScrollerImpSPI.h:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::scrollbarColorStyle const):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::setScrollbarColor):
(WebCore::AsyncScrollingCoordinator::setFrameScrollingNodeState):
(WebCore::AsyncScrollingCoordinator::setScrollingNodeScrollableAreaGeometry):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
* Source/WebCore/page/scrolling/ScrollingCoordinator.cpp:
(WebCore::ScrollingCoordinator::setScrollbarColor):
* Source/WebCore/page/scrolling/ScrollingCoordinator.h:
* Source/WebCore/page/scrolling/ScrollingStateFrameScrollingNode.cpp:
(WebCore::ScrollingStateFrameScrollingNode::ScrollingStateFrameScrollingNode):
* Source/WebCore/page/scrolling/ScrollingStateFrameScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingStateNode.h:
* Source/WebCore/page/scrolling/ScrollingStateOverflowScrollingNode.cpp:
(WebCore::ScrollingStateOverflowScrollingNode::ScrollingStateOverflowScrollingNode):
* Source/WebCore/page/scrolling/ScrollingStateOverflowScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingStatePluginScrollingNode.cpp:
(WebCore::ScrollingStatePluginScrollingNode::ScrollingStatePluginScrollingNode):
* Source/WebCore/page/scrolling/ScrollingStatePluginScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp:
(WebCore::ScrollingStateScrollingNode::ScrollingStateScrollingNode):
(WebCore::ScrollingStateScrollingNode::setScrollbarColor):
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.h:
(WebCore::ScrollingStateScrollingNode::scrollbarColor const):
* Source/WebCore/page/scrolling/mac/ScrollerMac.h:
* Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
(WebCore::ScrollerMac::updateValues):
(WebCore::ScrollerMac::scrollbarColorChanged):
(WebCore::ScrollerMac::scrollbarState const):
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.h:
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
(WebCore::ScrollerPairMac::scrollbarColorChanged):
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
(WebCore::ScrollingTreeScrollingNodeDelegateMac::updateFromStateNode):
* Source/WebCore/platform/ScrollTypes.h:
* Source/WebCore/platform/ScrollableArea.cpp:
(WebCore::ScrollableArea::scrollbarColorDidChange):
* Source/WebCore/platform/ScrollableArea.h:
(WebCore::ScrollableArea::scrollbarColorStyle const):
* Source/WebCore/platform/ScrollbarsController.cpp:
(WebCore::ScrollbarsController::scrollbarColorChanged):
* Source/WebCore/platform/ScrollbarsController.h:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::styleDidChange):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::scrollbarColorStyle const):
* Source/WebCore/rendering/RenderLayerScrollableArea.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.mm:
(WebKit::RemoteScrollbarsController::scrollbarColorChanged):

Canonical link: <a href="https://commits.webkit.org/298735@main">https://commits.webkit.org/298735@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9eefd1d43425902fe4f4bcad476817b6ab9e46b2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116491 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36152 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26704 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122548 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/67054 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118380 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36846 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44740 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88460 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42930 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119440 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29384 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104502 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68904 "") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28452 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22609 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66229 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/108601 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98759 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22769 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125697 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/115019 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43385 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32577 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97165 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43750 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100711 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96960 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24678 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42259 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20170 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/39339 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43273 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48864 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/143717 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42739 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37012 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46079 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44444 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

nt1m reviewed (commented):
I think `std::optional<std::pair<Color, Color>>` represents more accurately the property than `std::pair<std::optional<Color>, std::optional<Color>>` fwiw

nt1m commented on LayoutTests/platform/mac-sequoia/TestExpectations:
```diff
@@ -41,6 +41,9 @@ imported/w3c/web-platform-tests/css/css-ui/button-author-level-padding-applies.h
 fast/forms/datalist/input-appearance-range-with-datalist-rtl.html [ Pass ]
 
+# Test only passes with AppKit support for scrollbar color, added in Tahoe
+fast/scrolling/mac/scrollbars/scrollbar-color.html
```
```suggestion
fast/scrolling/mac/scrollbars/scrollbar-color.html [ Skip ]
```

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
```diff
@@ -526,6 +542,17 @@ - (void)invalidate
+    }
+
+    if ([m_scrollerImp knobColor] != nil) {
+        result.append(","_s);
```
nit: the logging should probably say which is which color

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -458,6 +458,17 @@ void AsyncScrollingCoordinator::setScrollbarLayoutDirection(ScrollableArea& scro
     stateNode->setScrollbarLayoutDirection(scrollbarLayoutDirection);
 }
 
+void AsyncScrollingCoordinator::setScrollbarColorState(ScrollableArea& scrollableArea, std::optional<std::pair<Color, Color>> scrollbarColorState)
```
Here and in the rest of the PR, could we drop "State" to be consistent with "ScrollbarWidth"?

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/page/scrolling/ScrollingStateScrollingNode.h:
```diff
@@ -55,6 +55,12 @@ struct ScrollbarEnabledState {
+    std::optional<Color> thumbColor;
+    std::optional<Color> trackColor;
+    friend bool operator==(const ScrollbarColorState&, const ScrollbarColorState&) = default;
+};
```
Could we make it this and use it as `std::optional<ScrollbarColor>` given we always pass all or no colors?
```suggestion
struct ScrollbarColor {
    Color thumbColor;
    Color trackColor;
    friend bool operator==(const ScrollbarColor&, const ScrollbarColor&) = default;
};
```

I'd even move the struct to `Source/WebCore/platform/` so we can use that struct instead of `std::pair`

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -6903,6 +6903,17 @@ ScrollbarWidth LocalFrameView::scrollbarWidthStyle()  const
     return ScrollbarWidth::Auto;
 }
 
+std::optional<std::pair<WebCore::Color, WebCore::Color>> LocalFrameView::scrollbarColorStyle() const
```
Looking at the full PR, we should probably just re-use the `ScrollbarColor` struct, which should let you write nice things like:

```
= {
   .thumb = ...,
   .track = ...,
}
```

nmoucht replied:
Fixed.

nt1m reviewed (commented):
Looks generally OK

nt1m commented on LayoutTests/fast/scrolling/mac/scrollbars/scrollbar-color.html:
```diff
@@ -0,0 +1,76 @@
+            height: 200%;
+        }
+    </style>
+    <script src="../../../../resources/js-test-pre.js"></script>
```
nit: please use js-test.js instead of pre/post

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -458,6 +458,17 @@ void AsyncScrollingCoordinator::setScrollbarLayoutDirection(ScrollableArea& scro
     stateNode->setScrollbarLayoutDirection(scrollbarLayoutDirection);
 }
 
+void AsyncScrollingCoordinator::setScrollbarColor(ScrollableArea& scrollableArea, std::optional<ScrollbarColor> scrollbarColorState)
```
```suggestion
void AsyncScrollingCoordinator::setScrollbarColor(ScrollableArea& scrollableArea, std::optional<ScrollbarColor> scrollbarColor)
```

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -458,6 +458,17 @@ void AsyncScrollingCoordinator::setScrollbarLayoutDirection(ScrollableArea& scro
+    if (!stateNode)
+        return;
+
+    stateNode->setScrollbarColor(scrollbarColorState);
```
ditto here

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
```diff
@@ -499,6 +503,18 @@ - (void)invalidate
+{
+    if (scrollbarColorState) {
+        m_trackColor = cocoaColor(scrollbarColorState->trackColor);
+        m_thumbColor = cocoaColor(scrollbarColorState->thumbColor);
```
```suggestion
void ScrollerMac::scrollbarColorChanged(const std::optional<ScrollbarColor>& scrollbarColor)
{
    if (scrollbarColor) {
        m_trackColor = cocoaColor(scrollbarColor->trackColor);
        m_thumbColor = cocoaColor(scrollbarColor->thumbColor);
```

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
```diff
@@ -444,6 +444,12 @@ - (void)scrollerImpPair:(NSScrollerImpPair *)scrollerImpPair updateScrollerStyle
+void ScrollerPairMac::scrollbarColorChanged(const std::optional<ScrollbarColor>& scrollbarColorState)
+{
+    checkedHorizontalScroller()->scrollbarColorChanged(scrollbarColorState);
+    checkedVerticalScroller()->scrollbarColorChanged(scrollbarColorState);
```
```suggestion
void ScrollerPairMac::scrollbarColorChanged(const std::optional<ScrollbarColor>& scrollbarColor)
{
    checkedHorizontalScroller()->scrollbarColorChanged(scrollbarColor);
    checkedVerticalScroller()->scrollbarColorChanged(scrollbarColor);
```

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -372,6 +372,18 @@ void RenderBox::styleDidChange(StyleDifference diff, const RenderStyle* oldStyle
+        std::optional<ScrollbarColor> scrollbarColor;
+
+        if (auto value = newStyle.scrollbarColor().tryValue())
+            scrollbarColor = { newStyle.colorResolvingCurrentColor(value->thumb), newStyle.colorResolvingCurrentColor(value->track) };
```
```suggestion
            scrollbarColor = {
                .thumb = newStyle.colorResolvingCurrentColor(value->thumb),
                .track = newStyle.colorResolvingCurrentColor(value->track)
            };
```

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
```diff
@@ -1064,6 +1064,16 @@ ScrollbarWidth RenderLayerScrollableArea::scrollbarWidthStyle()  const
+{
+    if (m_layer.renderBox()) {
+        if (auto value = m_layer.renderer().style().scrollbarColor().tryValue())
+            return ScrollbarColor { m_layer.renderer().style().colorResolvingCurrentColor(value->thumb), m_layer.renderer().style().colorResolvingCurrentColor(value->track) };
```
```suggestion
            return {
                .thumb = m_layer.renderer().style().colorResolvingCurrentColor(value->thumb),
                .track = m_layer.renderer().style().colorResolvingCurrentColor(value->track)
            };
```
(Maybe ScrollbarColor is necessary here because of the optional, not sure)

nmoucht replied:
Fixed.

nt1m commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.mm:
```diff
@@ -152,5 +152,12 @@
+{
+    if (auto scrollingCoordinator = m_coordinator.get())
+        scrollingCoordinator->setScrollbarColor(scrollableArea(), colorState);
+}
```
```suggestion
void RemoteScrollbarsController::scrollbarColorChanged(std::optional<WebCore::ScrollbarColor> color)
{
    if (auto scrollingCoordinator = m_coordinator.get())
        scrollingCoordinator->setScrollbarColor(scrollableArea(), color);
}
```

nmoucht replied:
Fixed.

smfr reviewed (approved):
Let's also add a test to test that `scrollbar-color` on the body doesn't propagate to the root (which from memory is the expected behavior).

nmoucht committed:
[MacOS] Implement scrollbar-color
https://bugs.webkit.org/show_bug.cgi?id=297166
rdar://157906557

Reviewed by Tim Nguyen and Simon Fraser.

Adopt new AppKit SPI for setting the color of the scrollbar thumb and track.

* LayoutTests/fast/scrolling/mac/scrollbars/scrollbar-color-body-expected.txt: Added.
* LayoutTests/fast/scrolling/mac/scrollbars/scrollbar-color-body.html: Added.
* LayoutTests/fast/scrolling/mac/scrollbars/scrollbar-color-expected.txt: Added.
* LayoutTests/fast/scrolling/mac/scrollbars/scrollbar-color.html: Added.
* LayoutTests/platform/mac-sequoia/TestExpectations:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/PAL/pal/spi/mac/NSScrollerImpSPI.h:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::scrollbarColorStyle const):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::setScrollbarColor):
(WebCore::AsyncScrollingCoordinator::setFrameScrollingNodeState):
(WebCore::AsyncScrollingCoordinator::setScrollingNodeScrollableAreaGeometry):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
* Source/WebCore/page/scrolling/ScrollingCoordinator.cpp:
(WebCore::ScrollingCoordinator::setScrollbarColor):
* Source/WebCore/page/scrolling/ScrollingCoordinator.h:
* Source/WebCore/page/scrolling/ScrollingStateFrameScrollingNode.cpp:
(WebCore::ScrollingStateFrameScrollingNode::ScrollingStateFrameScrollingNode):
* Source/WebCore/page/scrolling/ScrollingStateFrameScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingStateNode.h:
* Source/WebCore/page/scrolling/ScrollingStateOverflowScrollingNode.cpp:
(WebCore::ScrollingStateOverflowScrollingNode::ScrollingStateOverflowScrollingNode):
* Source/WebCore/page/scrolling/ScrollingStateOverflowScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingStatePluginScrollingNode.cpp:
(WebCore::ScrollingStatePluginScrollingNode::ScrollingStatePluginScrollingNode):
* Source/WebCore/page/scrolling/ScrollingStatePluginScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp:
(WebCore::ScrollingStateScrollingNode::ScrollingStateScrollingNode):
(WebCore::ScrollingStateScrollingNode::setScrollbarColor):
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.h:
(WebCore::ScrollingStateScrollingNode::scrollbarColor const):
* Source/WebCore/page/scrolling/mac/ScrollerMac.h:
* Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
(WebCore::ScrollerMac::updateValues):
(WebCore::ScrollerMac::scrollbarColorChanged):
(WebCore::ScrollerMac::scrollbarState const):
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.h:
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
(WebCore::ScrollerPairMac::scrollbarColorChanged):
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
(WebCore::ScrollingTreeScrollingNodeDelegateMac::updateFromStateNode):
* Source/WebCore/platform/ScrollTypes.h:
* Source/WebCore/platform/ScrollableArea.cpp:
(WebCore::ScrollableArea::scrollbarColorDidChange):
* Source/WebCore/platform/ScrollableArea.h:
(WebCore::ScrollableArea::scrollbarColorStyle const):
* Source/WebCore/platform/ScrollbarsController.cpp:
(WebCore::ScrollbarsController::scrollbarColorChanged):
* Source/WebCore/platform/ScrollbarsController.h:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::styleDidChange):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::scrollbarColorStyle const):
* Source/WebCore/rendering/RenderLayerScrollableArea.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.mm:
(WebKit::RemoteScrollbarsController::scrollbarColorChanged):

Canonical link: https://commits.webkit.org/298735@main
